cmd/kubeadm/app/util/runtime/runtime.go

```diff
@@ -38,64 +41,89 @@ var defaultKnownCRISockets = []string{
 
 // ContainerRuntime is an interface for working with container runtimes
 type ContainerRuntime interface {
-	Socket() string
+	Connect() error
+	SetImpl(impl)
 	IsRunning() error
 	ListKubeContainers() ([]string, error)
 	RemoveContainers(containers []string) error
 	PullImage(image string) error
 	PullImagesInParallel(images []string, ifNotPresent bool) error
-	ImageExists(image string) (bool, error)
+	ImageExists(image string) bool
 	SandboxImage() (string, error)
 }
 
 // CRIRuntime is a struct that interfaces with the CRI
 type CRIRuntime struct {
-	exec       utilsexec.Interface
-	criSocket  string
-	crictlPath string
+	impl           impl
+	logger         *klog.Logger
+	criSocket      string
+	runtimeService criapi.RuntimeService
+	imageService   criapi.ImageManagerService
 }
 
+// defaultTimeout is the default timeout inherited by crictl
+const defaultTimeout = 2 * time.Second
+
 // NewContainerRuntime sets up and returns a ContainerRuntime struct
-func NewContainerRuntime(execer utilsexec.Interface, criSocket string) (ContainerRuntime, error) {
-	const toolName = "crictl"
-	crictlPath, err := execer.LookPath(toolName)
-	if err != nil {
-		return nil, errors.Wrapf(err, "%s is required by the container runtime", toolName)
+func NewContainerRuntime(criSocket string) ContainerRuntime {
+	logger := klog.Background()
+	return &CRIRuntime{
+		impl:      &defaultImpl{},
+		logger:    &logger,
+		criSocket: criSocket,
 	}
-	return &CRIRuntime{execer, criSocket, crictlPath}, nil
 }
 
-// Socket returns the CRI socket endpoint
-func (runtime *CRIRuntime) Socket() string {
-	return runtime.criSocket
+// SetImpl can be used to set the internal implementation for testing.
+func (runtime *CRIRuntime) SetImpl(impl impl) {
+	runtime.impl = impl
 }
 
-// crictl creates a crictl command for the provided args.
-func (runtime *CRIRuntime) crictl(args ...string) utilsexec.Cmd {
-	cmd := runtime.exec.Command(runtime.crictlPath, append([]string{"-r", runtime.Socket(), "-i", runtime.Socket()}, args...)...)
-	cmd.SetEnv(os.Environ())
-	return cmd
+// Connect establishes a connection with the CRI runtime.
+func (runtime *CRIRuntime) Connect() error {
+	runtimeService, err := runtime.impl.NewRemoteRuntimeService(runtime.criSocket, defaultTimeout, nil, runtime.logger)
+	if err != nil {
+		return fmt.Errorf("create new runtime service: %w", err)
+	}
+	runtime.runtimeService = runtimeService
+
+	imageService, err := runtime.impl.NewRemoteImageService(runtime.criSocket, defaultTimeout, nil, runtime.logger)
+	if err != nil {
+		return fmt.Errorf("create new image service: %w", err)
+	}
+	runtime.imageService = imageService
+
+	return nil
+}
+
+func defaultContext() (context.Context, context.CancelFunc) {
```

Comment:
i'd move this after the exported functions of the interface, maybe above pullImagesInParallelImpl